app-android/src/androidTest/kotlin/care/data4life/integration/app/crud/BaseCrudSdkTest.kt

```diff
@@ -0,0 +1,224 @@
+/*
+ * Copyright (c) 2022 D4L data4life gGmbH - All rights reserved.
+ */
+
+package care.data4life.integration.app.crud
+
+import care.data4life.integration.app.data.wrapper.Result
+import care.data4life.integration.app.data.wrapper.Result.Failure
+import care.data4life.integration.app.data.wrapper.Result.Success
+import care.data4life.integration.app.data.wrapper.awaitLegacyListener
+import care.data4life.integration.app.test.BaseSdkTest
+import care.data4life.sdk.call.CallContract.Record
+import org.junit.jupiter.api.Assertions
+import org.junit.jupiter.api.Assertions.fail
+import kotlin.test.assertEquals
+import care.data4life.sdk.model.DeleteResult as LegacyDeleteResult
+
+abstract class BaseCrudSdkTest<T> : BaseSdkTest() {
+
+    protected suspend fun runTestSuite(itemCount: Int) {
+        cleanAccount()
+
+        val initialItems = generateItems(itemCount)
+
+        val createdRecords = create(initialItems)
+
+        val createdItems = createdRecords.map { it.resource }
+        val recordIds = createdRecords.map { getId(it)!! }
+
+        count(itemCount)
+        fetch(recordIds, createdItems)
+        val loadedItems = fetchByType(itemCount, createdItems)
+
+        val mutatedItems = mutateItems(loadedItems)
+
+        update(recordIds, mutatedItems)
+        if (supportsDownload()) {
+            download(recordIds, mutatedItems)
+        }
+
+        delete(recordIds)
+        count(0)
+    }
+
+    private suspend fun create(items: List<T>): List<Record<T>> {
+        // Given
+        val createdRecords = mutableListOf<Record<T>>()
+
+        // When
+        val results: List<Result<Record<T>>> = callCreate(items)
+
+        // Then
+        assertResults(results, "Create") { index, record ->
+            assertCreateRecord(items[index], record)
+            createdRecords.add(record)
+        }
+
+        return createdRecords
+    }
+
+    abstract suspend fun callCreate(items: List<T>): List<Result<Record<T>>>
+    abstract fun assertCreateRecord(expected: T, actual: Record<T>)
+
+    private suspend fun count(count: Int) {
+        // Given
+
+        // When
+        val result: Result<Int> = callCount()
+
+        // Then
+        when (result) {
+            is Success -> {
+                Assertions.assertEquals(count, result.data)
+            }
+            is Failure -> {
+                fail("Count records failed: ${result.exception.message}")
+            }
```

Comment:
If you restructure the Result you could spare the when. To get [inspiration](https://github.com/bitPogo/pixalb/blob/bd1c9a26d84eca3b9b76ada839438e3e6f7c0c99/coroutine-helper/src/commonMain/kotlin/io/bitpogo/util/coroutine/result/Result.kt#L9)

Reply:
See above